src/persist-client/src/internal/sim.rs

```diff
@@ -0,0 +1,165 @@
+                None
+            }
+            Entry::Occupied(mut x) => {
+                let prev_idx = *x.get();
+                *x.get_mut() = idx;
+
+                let necessary_size = self.weights.range(prev_idx..).map(|(_, v)| *v).sum::<u64>();
+                assert!(self.weights.remove(&prev_idx).is_some());
+                assert!(self.weights.insert(idx, weight).is_none());
+                Some(necessary_size)
+            }
+        }
+    }
+}
+
+/// A [Blob] delegate that simulates the benefits of caching at various sizes.
+#[derive(Debug)]
+pub struct BlobCacheSim {
+    sim: Mutex<CacheSim<u64>>,
+    blob: Arc<dyn Blob + Send + Sync>,
+    metrics: Arc<Metrics>,
+}
+
+impl BlobCacheSim {
+    /// Returns a new BlobCacheSim delegating to the given Blob.
+    pub fn new(
+        blob: Arc<dyn Blob + Send + Sync>,
+        metrics: Arc<Metrics>,
+    ) -> Arc<dyn Blob + Send + Sync> {
+        let blob = BlobCacheSim {
+            sim: Default::default(),
+            blob,
+            metrics,
+        };
+        Arc::new(blob)
+    }
+
+    fn sim_key(key: &str) -> u64 {
+        let mut h = DefaultHasher::new();
+        key.hash(&mut h);
+        h.finish()
+    }
+}
+
+#[async_trait]
+impl Blob for BlobCacheSim {
+    async fn get(&self, key: &str) -> Result<Option<SegmentedBytes>, ExternalError> {
+        let blob = self.blob.get(key).await?;
+        if let Some(blob) = blob.as_ref() {
+            let sim_key = Self::sim_key(key);
+            let necessary_size = self
+                .sim
+                .lock()
+                .expect("lock")
+                .insert(sim_key, u64::cast_from(blob.len()));
+            if let Some(necessary_size) = necessary_size {
+                self.metrics
+                    .cache_sim
+                    .cached_blobs
+                    .observe(f64::cast_lossy(necessary_size));
+                self.metrics
+                    .cache_sim
+                    .observe_bytes(u64::cast_from(blob.len()), necessary_size)
+            }
+        }
+        Ok(blob)
+    }
+
+    async fn list_keys_and_metadata(
+        &self,
+        key_prefix: &str,
+        f: &mut (dyn FnMut(BlobMetadata) + Send + Sync),
+    ) -> Result<(), ExternalError> {
+        self.blob.list_keys_and_metadata(key_prefix, f).await
+    }
+
+    async fn set(&self, key: &str, value: Bytes, atomic: Atomicity) -> Result<(), ExternalError> {
+        let value_len = value.len();
+        let () = self.blob.set(key, value, atomic).await?;
+        let sim_key = Self::sim_key(key);
+        let _ = self
+            .sim
+            .lock()
+            .expect("lock")
+            .insert(sim_key, u64::cast_from(value_len));
+        Ok(())
+    }
+
+    async fn delete(&self, key: &str) -> Result<Option<usize>, ExternalError> {
+        // WIP delete key from sim
+        self.blob.delete(key).await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
```

Comment:
The linter will complain about this once https://github.com/MaterializeInc/materialize/pull/19603 lands, please rebase your change then and use `#[mz_ore::test]`.
```suggestion
    #[mz_ore::test]
```